plugins/flashrom/fu-plugin-flashrom.c

```diff
@@ -0,0 +1,198 @@
+};
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_free (data->flashrom_fn);
+}
+
+
+static void
+fu_plugin_flashrom_read_cb (const gchar *line, gpointer user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN (user_data);
+	if (g_strcmp0 (line, "Reading flash...") == 0)
+		fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_VERIFY);
+	//FIXME get percentage completion...
+	g_debug ("got '%s'", line);
+}
+
+static void
+fu_plugin_flashrom_write_cb (const gchar *line, gpointer user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN (user_data);
+	if (g_strcmp0 (line, "Writing flash...") == 0)
+		fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_WRITE);
+	//FIXME get percentage completion...
+	g_debug ("got '%s'", line);
+}
+
+static FuDevice *
+fu_plugin_flashrom_create_device (FuPlugin *plugin,
+				  const gchar *guid,
+				  const gchar *id,
+				  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autofree gchar *device_id = g_strdup_printf ("flashrom-%s", id);
+	g_autofree gchar *firmware_orig = NULL;
+	g_autofree gchar *basename = NULL;
+	g_autoptr(FuDevice) dev = fu_device_new ();
+
+	/* if the original firmware doesn't exist, grab it now */
+	basename = g_strdup_printf ("flashrom-%s.bin", id);
+	firmware_orig = g_build_filename (LOCALSTATEDIR, "lib", "fwupd",
+					  "builder", basename, NULL);
+	if (!fu_common_mkdir_parent (firmware_orig, error))
+		return FALSE;
+	if (!g_file_test (firmware_orig, G_FILE_TEST_EXISTS)) {
+		const gchar *argv[] = {
+			data->flashrom_fn,
+			"--programmer", "internal:laptop=force_I_want_a_brick",
+			"--read", firmware_orig,
+			"--verbose", NULL };
+		if (!fu_common_spawn_sync ((const gchar * const *) argv,
+					   fu_plugin_flashrom_read_cb, plugin,
+					   NULL, error)) {
+			g_prefix_error (error, "failed to get original firmware: ");
+			return NULL;
+		}
+		fu_plugin_set_status (plugin, FWUPD_STATUS_IDLE);
+	}
+
+	/* create device */
+	fu_device_set_id (dev, device_id);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_guid (dev, guid);
+	fu_device_set_version (dev, fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_BIOS_VERSION));
+	fu_device_set_name (dev, fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_PRODUCT_NAME));
+	fu_device_set_vendor (dev, fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_MANUFACTURER));
+	return g_steal_pointer (&dev);
+}
+
+gboolean
+fu_plugin_update (FuPlugin *plugin,
+		  FuDevice *device,
+		  GBytes *blob_fw,
+		  FwupdInstallFlags flags,
+		  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autofree gchar *flashrom_fn = NULL;
+	g_autofree gchar *firmware_fn = NULL;
+	g_autofree gchar *tmpdir = NULL;
+	const gchar *argv[] = {
+		"/xxx/flashrom",
+		"--programmer", "internal:laptop=force_I_want_a_brick",
```

Comment:
Likewise is this really the right option?

Reply:
It is, for laptops at least. Maybe in the future we could put the programmer into the quirk details, but not yet.